File: src/cogs/Betting.py
from discord.ext import commands
import discord

from src.database.balance import get_balance, update_balance
from src.database.bet import create_bet_db, get_bet_data, add_wager, close_bet_db, freeze_bet
from src.database.achievement import increment_stat, check_and_unlock_achievements, format_achievements_unlocks
from src.database.settings import get_language
from src.utils.i18n import t


class Betting(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @commands.command(name='createbet')
    async def create_bet(self, ctx, description: str, option1: str, option2: str):
        """Créer un pari personnalisé."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        user_id = ctx.author.id
        bet_id = create_bet_db(user_id, description, option1, option2)
        await ctx.send(t("betting.created", lang, id=bet_id, desc=description))

    @commands.command(name='bet')
    async def place_bet(self, ctx, bet_id: str, choice: str, amount: int):
        """Parier sur un pari personnalisé."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        choice = choice.lower()
        bet_data = get_bet_data(bet_id)
        if not bet_data:
            return await ctx.send(t("betting.not_found", lang, id=bet_id))
        if bet_data["status"] == "CLOSE":
            return await ctx.send(t("betting.finished", lang))
        if bet_data["status"] == "FROZEN":
            return await ctx.send(t("betting.frozen", lang))
        current_balance = get_balance(ctx.author.id)
        if current_balance < amount:
            return await ctx.send(t("betting.no_money", lang, balance=current_balance))
        update_balance(ctx.author.id, -amount)
        choice_description = bet_data["options"][0] if choice == "a" else bet_data["options"][1]
        add_wager(bet_id, ctx.author.id, choice, amount)
        return await ctx.send(t("betting.placed", lang, choice=choice_description))

    @commands.command(name='closebet')
    async def close_bet(self, ctx, bet_id: str, winning_option: str):
        """Terminer un pari et distribuer les gains."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        bet_data = get_bet_data(bet_id)
        winning_option = winning_option.lower()
        if not bet_data:
            return await ctx.send(t("betting.id_not_found", lang))
        bet_id = bet_data["id"]
        if ctx.author.id != bet_data["creator"]:
            return await ctx.send(t("betting.only_creator", lang))
        if bet_data["status"] == "CLOSE":
            return await ctx.send(t("betting.already_closed", lang))
        if winning_option not in ["a", "b"]:
            return await ctx.send(t("betting.invalid_choice", lang))
            
        bet_winning_option = bet_data['options'][0] if winning_option == "a" else bet_data['options'][1]
        total_pool = sum(w["amount"] for w in bet_data["wagers"])
        winning_pool = sum(w["amount"] for w in bet_data["wagers"] if w["option"] == winning_option)
        results = []
        if winning_pool == 0:
            await ctx.send(t("betting.closed_house_keeps", lang, option=bet_winning_option, pool=total_pool))
        else:
            multiplier = total_pool / winning_pool
            for wager in bet_data["wagers"]:
                user_id = wager["user_id"]
                if wager["option"] == winning_option:
                    increment_stat(user_id, "wagers_won")
                    wager_amount = wager["amount"]
                    payout = int(wager_amount * multiplier)
                    update_balance(user_id, payout)
                    user = self.bot.get_user(user_id)
                    name = user.display_name if user else "Unknown"
                    results.append(t("betting.won_msg", lang, user=name, amount=payout))
                else:
                    increment_stat(user_id, "wagers_lost")
                
                unlocks = check_and_unlock_achievements(user_id)
                if unlocks:
                    user_obj = self.bot.get_user(user_id)
                    if user_obj:
                        await ctx.send(content=user_obj.mention, embed=format_achievements_unlocks(unlocks, lang))
            
            embed = discord.Embed(title=t("betting.result_title", lang, id=bet_id), description=f"{t('item_manager.winner', lang)}: **{bet_winning_option}**",
                                  color=discord.Color.purple())
            embed.add_field(name=t("betting.total_pool", lang), value=f"${total_pool}")
            embed.add_field(name=t("betting.winners", lang), value="\n".join(results) if results else "None")
            await ctx.send(embed=embed)
        close_bet_db(bet_id, winning_option)
        return None

    @commands.command(name='odds', aliases=['betinfo', 'status'])
    async def show_odds(self, ctx, bet_id: str):
        """Voir les cotes et infos d'un pari."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        bet_data = get_bet_data(bet_id)
        if not bet_data:
            return await ctx.send(t("inventory.item_not_found", lang).replace(t("inventory.item", lang), "bet"))
        
        bet_id = bet_data["id"]
        total_pool = sum(w["amount"] for w in bet_data["wagers"])
        opt1 = bet_data["options"][0]
        opt2 = bet_data["options"][1]
        pool_1 = sum(w["amount"] for w in bet_data["wagers"] if w["option"] == "a")
        pool_2 = sum(w["amount"] for w in bet_data["wagers"] if w["option"] == "b")
        odds_1 = round(total_pool / pool_1, 2) if pool_1 > 0 else "N/A"
        odds_2 = round(total_pool / pool_2, 2) if pool_2 > 0 else "N/A"

        embed = discord.Embed(title=t("betting.status_title", lang, id=bet_id), description=bet_data["description"],
                               color=discord.Color.blue())
        embed.add_field(name=t("betting.total_bet", lang), value=f"${total_pool}", inline=False)
        embed.add_field(
            name=t("betting.option_a", lang, name=opt1.capitalize()),
            value=f"**Value:** ${pool_1}\n**Odds:** {odds_1}x",
            inline=True
        )
        embed.add_field(
            name=t("betting.option_b", lang, name=opt2.capitalize()),
            value=f"**Value:** ${pool_2}\n**Odds:** {odds_2}x",
            inline=True
        )
        if bet_data["status"] == "CLOSED" or bet_data["status"] == "CLOSE":
            embed.set_footer(text=t("betting.odds_footer_closed", lang, winner=bet_data.get('winner', 'Unknown')))
        else:
            embed.set_footer(text=t("betting.odds_footer_open", lang))
        return await ctx.send(embed=embed)

    @commands.command(name="freezebet")
    async def freeze_bet(self, ctx, bet_id: str):
        """Geler un pari pour empêcher de nouvelles mises."""
        lang = get_language(ctx.guild.id if ctx.guild else None)
        bet_data = get_bet_data(bet_id)
        if not bet_data:
            return await ctx.send(t("betting.id_not_found", lang))
        if ctx.author.id != bet_data["creator"]:
            return await ctx.send(t("betting.only_creator", lang))
        if bet_data["status"] == "CLOSE":
            return await ctx.send(t("betting.already_closed", lang))
        if bet_data["status"] == "FROZEN":
            return await ctx.send(t("betting.frozen", lang))
        freeze_bet(bet_id)
        return await ctx.send(t("betting.freeze_success", lang, desc=bet_data["description"]))


async def setup(bot):
    await bot.add_cog(Betting(bot))

File: internal/cogs/betting/betting.go
package betting

import (
	"strconv"
	"strings"

	"github.com/bwmarrin/discordgo"

	"guacagamblebot/internal/components"
	"guacagamblebot/internal/config"
	"guacagamblebot/internal/i18n"
	"guacagamblebot/internal/interaction"
	bettingsvc "guacagamblebot/internal/service/betting"
	"guacagamblebot/internal/store"
)

type Cog struct {
	store *store.Store
	cfg   *config.Config
	svc   *bettingsvc.Service
}

func Register(r *interaction.Router, s *store.Store, cfg *config.Config) {
	c := &Cog{store: s, cfg: cfg, svc: bettingsvc.New(s, cfg)}
	r.Slash("betting", "Paris personnalisés : créer, parier, clôturer.", c.onSlashMenu)
	r.Prefix("betting", c.onPrefixMenu)
	r.Component("betting", "create", c.onCreateOpen)
	r.Component("betting", "place", c.onPlaceOpen)
	r.Component("betting", "close", c.onCloseOpen)
	r.Component("betting", "odds", c.onOddsOpen)
	r.Component("betting", "freeze", c.onFreezeOpen)
	r.Modal("betting", "create_submit", c.onCreateSubmit)
	r.Modal("betting", "place_submit", c.onPlaceSubmit)
	r.Modal("betting", "close_submit", c.onCloseSubmit)
	r.Modal("betting", "odds_submit", c.onOddsSubmit)
	r.Modal("betting", "freeze_submit", c.onFreezeSubmit)
}

func (c *Cog) onSlashMenu(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	embed, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseChannelMessageWithSource, embed, comps))
}

func (c *Cog) onPrefixMenu(b *interaction.Bot, s *discordgo.Session, m *discordgo.Message) {
	lang := c.store.GetLanguage(interaction.ToInt64(m.GuildID))
	embed, comps := c.menu(lang)
	_, _ = s.ChannelMessageSendComplex(m.ChannelID, &discordgo.MessageSend{
		Embeds:     []*discordgo.MessageEmbed{embed},
		Components: comps,
	})
}

func (c *Cog) menu(lang string) (*discordgo.MessageEmbed, []discordgo.MessageComponent) {
	embed := components.Embed(
		i18n.T("betting.status_title", lang, map[string]any{"id": "?"}),
		i18n.T("betting.odds_footer_open", lang),
		0x9b59b6,
	)
	comps := []discordgo.MessageComponent{
		components.ActionRow(
			components.Button("🎲 "+i18n.T("item_manager.accept_label", lang), components.Encode("betting", "create"), discordgo.SuccessButton),
			components.Button("🎯 "+i18n.T("betting.won_msg", lang, map[string]any{"user": "", "amount": ""}), components.Encode("betting", "place"), discordgo.PrimaryButton),
		),
		components.ActionRow(
			components.Button("📊 "+i18n.T("betting.total_bet", lang), components.Encode("betting", "odds"), discordgo.SecondaryButton),
			components.Button("🔒 "+i18n.T("betting.result_title", lang, map[string]any{"id": ""}), components.Encode("betting", "close"), discordgo.DangerButton),
			components.Button("❄️ "+i18n.T("betting.finished", lang), components.Encode("betting", "freeze"), discordgo.SecondaryButton),
		),
	}
	return embed, comps
}

func (c *Cog) onCreateOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("betting", "create_submit"),
		i18n.T("betting.odds_footer_open", lang),
		components.TextInput("description", "Description", true, "Who will win?", discordgo.TextInputShort, 1, 100),
		components.TextInput("option1", "Option A", true, "Team A", discordgo.TextInputShort, 1, 50),
		components.TextInput("option2", "Option B", true, "Team B", discordgo.TextInputShort, 1, 50),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onPlaceOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("betting", "place_submit"),
		i18n.T("betting.odds_footer_open", lang),
		components.TextInput("bet_id", "Bet ID", true, "1", discordgo.TextInputShort, 1, 12),
		components.TextInput("option", "a or b", true, "a", discordgo.TextInputShort, 1, 1),
		components.TextInput("amount", i18n.T("economy.quantity", lang), true, "100", discordgo.TextInputShort, 1, 12),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onCloseOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("betting", "close_submit"),
		i18n.T("betting.result_title", lang, map[string]any{"id": ""}),
		components.TextInput("bet_id", "Bet ID", true, "1", discordgo.TextInputShort, 1, 12),
		components.TextInput("winner", "Winning option (a or b)", true, "a", discordgo.TextInputShort, 1, 1),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onOddsOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("betting", "odds_submit"),
		i18n.T("betting.total_bet", lang),
		components.TextInput("bet_id", "Bet ID", true, "1", discordgo.TextInputShort, 1, 12),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onFreezeOpen(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	modal := components.ModalResponse(
		components.Encode("betting", "freeze_submit"),
		i18n.T("betting.finished", lang),
		components.TextInput("bet_id", "Bet ID", true, "1", discordgo.TextInputShort, 1, 12),
	)
	_ = b.Session.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseModal,
		Data: modal,
	})
}

func (c *Cog) onCreateSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	userID := interaction.ToInt64(interaction.UserID(i))
	values := interaction.ModalValues(i)

	desc := strings.TrimSpace(values["description"])
	opt1 := strings.TrimSpace(values["option1"])
	opt2 := strings.TrimSpace(values["option2"])
	if desc == "" || opt1 == "" || opt2 == "" {
		interaction.RespondError(b, i, lang, "betting.id_not_found")
		return
	}

	betID, err := c.svc.CreateBet(userID, desc, opt1, opt2)
	if err != nil {
		interaction.RespondError(b, i, lang, "betting.id_not_found")
		return
	}

	embed := components.Embed(
		i18n.T("betting.created", lang, map[string]any{"id": betID, "desc": desc}),
		"",
		0x2ecc71,
	)
	_, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}

func (c *Cog) onPlaceSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	userID := interaction.ToInt64(interaction.UserID(i))
	values := interaction.ModalValues(i)

	betID, err := strconv.ParseInt(strings.TrimSpace(values["bet_id"]), 10, 64)
	if err != nil {
		interaction.RespondError(b, i, lang, "betting.not_found")
		return
	}

	choice := strings.ToLower(strings.TrimSpace(values["option"]))
	amount, aerr := strconv.Atoi(strings.TrimSpace(values["amount"]))
	if aerr != nil || amount <= 0 {
		interaction.RespondError(b, i, lang, "betting.no_money")
		return
	}

	perr := c.svc.PlaceBet(userID, betID, choice, amount)
	if perr != nil {
		switch perr {
		case bettingsvc.ErrNotFound:
			interaction.RespondError(b, i, lang, "betting.not_found")
		case bettingsvc.ErrClosed:
			interaction.RespondError(b, i, lang, "betting.finished")
		case bettingsvc.ErrFrozen:
			interaction.RespondError(b, i, lang, "betting.frozen")
		case bettingsvc.ErrNoMoney:
			interaction.RespondError(b, i, lang, "betting.no_money")
		default:
			interaction.RespondError(b, i, lang, "betting.not_found")
		}
		return
	}

	embed := components.Embed(
		i18n.T("betting.placed", lang, map[string]any{"choice": choice}),
		"",
		0x2ecc71,
	)
	_, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}

func (c *Cog) onCloseSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	userID := interaction.ToInt64(interaction.UserID(i))
	values := interaction.ModalValues(i)

	betID, err := strconv.ParseInt(strings.TrimSpace(values["bet_id"]), 10, 64)
	if err != nil {
		interaction.RespondError(b, i, lang, "betting.id_not_found")
		return
	}

	winner := strings.ToLower(strings.TrimSpace(values["winner"]))

	res, cerr := c.svc.CloseBet(userID, betID, winner)
	if cerr != nil {
		switch cerr {
		case bettingsvc.ErrNotFound:
			interaction.RespondError(b, i, lang, "betting.id_not_found")
		case bettingsvc.ErrNotCreator:
			interaction.RespondError(b, i, lang, "betting.only_creator")
		case bettingsvc.ErrClosed:
			interaction.RespondError(b, i, lang, "betting.already_closed")
		case bettingsvc.ErrInvalidOpt:
			interaction.RespondError(b, i, lang, "betting.invalid_choice")
		default:
			interaction.RespondError(b, i, lang, "betting.id_not_found")
		}
		return
	}

	winningOptName := "A"
	if winner == "b" {
		winningOptName = "B"
	}

	desc := i18n.T("item_manager.winner", lang) + ": **" + winningOptName + "**"
	if res.WinningPool == 0 {
		desc = i18n.T("betting.closed_house_keeps", lang, map[string]any{
			"option": winningOptName,
			"pool":   res.TotalPool,
		})
	}

	embed := components.Embed(
		i18n.T("betting.result_title", lang, map[string]any{"id": betID}),
		desc,
		0x9b59b6,
	)

	if len(res.WagerResults) > 0 {
		var winners []string
		for _, wr := range res.WagerResults {
			if wr.Won {
				winners = append(winners, i18n.T("betting.won_msg", lang, map[string]any{
					"user":   interaction.Mention(wr.UserID),
					"amount": wr.Amount,
				}))
			}
		}
		if len(winners) > 0 {
			embed.Fields = []*discordgo.MessageEmbedField{
				components.Field(i18n.T("betting.total_pool", lang), "$"+strconv.Itoa(res.TotalPool), false),
				components.Field(i18n.T("betting.winners", lang), strings.Join(winners, "\n"), false),
			}
		}
	}

	_, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}

func (c *Cog) onOddsSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	values := interaction.ModalValues(i)

	betID, err := strconv.ParseInt(strings.TrimSpace(values["bet_id"]), 10, 64)
	if err != nil {
		interaction.RespondError(b, i, lang, "betting.not_found")
		return
	}

	odds, oerr := c.svc.ShowOdds(betID)
	if oerr != nil {
		interaction.RespondError(b, i, lang, "betting.not_found")
		return
	}

	embed := components.Embed(
		i18n.T("betting.status_title", lang, map[string]any{"id": odds.BetID}),
		odds.Description,
		0x3498db,
	)
	embed.Fields = []*discordgo.MessageEmbedField{
		components.Field(i18n.T("betting.total_bet", lang), "$"+strconv.Itoa(odds.Total), false),
		components.Field(i18n.T("betting.option_a", lang, map[string]any{"name": odds.Option1}), "$"+strconv.Itoa(odds.Pool1)+" | "+odds.Odds1, true),
		components.Field(i18n.T("betting.option_b", lang, map[string]any{"name": odds.Option2}), "$"+strconv.Itoa(odds.Pool2)+" | "+odds.Odds2, true),
	}

	if odds.Status == "CLOSE" || odds.Status == "FROZEN" {
		embed.Footer = &discordgo.MessageEmbedFooter{Text: i18n.T("betting.odds_footer_closed", lang, map[string]any{"winner": odds.Winner})}
	} else {
		embed.Footer = &discordgo.MessageEmbedFooter{Text: i18n.T("betting.odds_footer_open", lang)}
	}

	_, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}

func (c *Cog) onFreezeSubmit(b *interaction.Bot, i *discordgo.InteractionCreate) {
	lang := c.store.GetLanguage(interaction.ToInt64(i.GuildID))
	userID := interaction.ToInt64(interaction.UserID(i))
	values := interaction.ModalValues(i)

	betID, err := strconv.ParseInt(strings.TrimSpace(values["bet_id"]), 10, 64)
	if err != nil {
		interaction.RespondError(b, i, lang, "betting.id_not_found")
		return
	}

	ferr := c.svc.FreezeBet(userID, betID)
	if ferr != nil {
		switch ferr {
		case bettingsvc.ErrNotFound:
			interaction.RespondError(b, i, lang, "betting.id_not_found")
		case bettingsvc.ErrNotCreator:
			interaction.RespondError(b, i, lang, "betting.only_creator")
		case bettingsvc.ErrClosed:
			interaction.RespondError(b, i, lang, "betting.already_closed")
		case bettingsvc.ErrFrozen:
			interaction.RespondError(b, i, lang, "betting.frozen")
		default:
			interaction.RespondError(b, i, lang, "betting.id_not_found")
		}
		return
	}

	embed := components.Embed(
		i18n.T("betting.freeze_success", lang, map[string]any{"desc": "Bet #" + strconv.FormatInt(betID, 10)}),
		"",
		0x2ecc71,
	)
	_, comps := c.menu(lang)
	_ = b.Session.InteractionRespond(i.Interaction,
		components.InteractionResponse(discordgo.InteractionResponseUpdateMessage, embed, comps))
}
